torch/csrc/distributed/c10d/init.cpp

```diff
@@ -51,6 +51,25 @@
 
 namespace {
 
+#ifdef USE_C10D_NCCL
+
+bool acquire_gil() {
+  // basically if this function can acquire the gil, it will return quickly.
+  // if not, it will hang forever.  The idea is to call this from a thread
+  // wrapped in a future, and then check the future after a timeout, to
+  // determine whether we're facing gil contention.
+  pybind11::gil_scoped_acquire gil;
+  return true;
+}
+
+bool registerGilChecker() {
+  c10d::get_gil_checker() = &acquire_gil;
```

Comment:
We might want to put this is. `Py_IsInitialized()` check. It isn't the current problem, but can we guarantee Python isn't exiting when GIL checker gets run?

Reply:
i suppose the main thread can exit and destroy python while the watchdog is still alive and hanging, and in the (probably unlikely) race that the heartbeat monitor notices the hanging watchdog before the whole process tree goes away, we'd hit this condition.

also, it seems like we have to make the (unsafe?) assumption that if Py_IsInitialized() returns true, it won't return false moments later when we actually try to acquire the gil.